src/main/java/org/broadinstitute/hellbender/tools/spark/ParallelCopyGCSDirectoryIntoHDFSSpark.java

```diff
@@ -0,0 +1,171 @@
+                }
+            }
+
+            if (chunkList.size() == 0) {
+                logger.info("no files found to copy");
+                return;
+            }
+
+            final JavaPairRDD<String, Integer> chunkRDD = ctx.parallelizePairs(chunkList, chunkList.size());
+
+            final JavaPairRDD<String, Tuple2<Integer, String>> chunkMappingRDD =
+                    chunkRDD.mapToPair(p -> new Tuple2<>(p._1(), readChunkToHdfs(auth, p._1(), chunkSize, p._2(), outputDirectoryFinal)));
+
+            final Map<String, Iterable<Tuple2<Integer, String>>> chunksByFilePath = chunkMappingRDD.groupByKey().collectAsMap();
+
+            for (Path path : gcsPaths) {
+
+                if (Files.isDirectory(path)) {
+                    continue;
+                }
+
+                final String filePath = path.toUri().toString();
+                final Iterable<Tuple2<Integer, String>> chunkListForFile = chunksByFilePath.get(filePath);
+                final String basename = path.getName(path.getNameCount() - 1).toString();
+                final org.apache.hadoop.fs.Path outFilePath = new org.apache.hadoop.fs.Path(outputDirectoryFinal + "/" + basename);
+                fs.createNewFile(outFilePath);
+
+                SortedMap<Integer, String> chunkMap = new TreeMap<>();
+                for (Tuple2<Integer, String> entry : chunkListForFile) {
+                    chunkMap.put(entry._1(), entry._2());
+                }
+
+                org.apache.hadoop.fs.Path[] chunkPaths = new org.apache.hadoop.fs.Path[chunkMap.size()];
+
+                final Iterator<Integer> iterator = chunkMap.keySet().iterator();
+                while (iterator.hasNext()) {
+                    final Integer next =  iterator.next();
+                    final String chunkPath = chunkMap.get(next);
+                    chunkPaths[next] = new org.apache.hadoop.fs.Path(chunkPath);
+                }
+
+                fs.concat(outFilePath, chunkPaths);
+            }
+
+
+
+        } catch (IOException e) {
+            throw new GATKException(e.getMessage(), e);
+        }
+
+    }
+
+    private static final Tuple2<Integer, String> readChunkToHdfs(final AuthHolder authHolder, final String inputGCSPathFinal, final long chunkSize, final Integer chunkNum, final String outputDirectory) {
+        final Path gcsPath = IOUtils.getPath(inputGCSPathFinal);
+        final String basename = gcsPath.getName(gcsPath.getNameCount() - 1).toString();
+        org.apache.hadoop.fs.Path outputPath = new org.apache.hadoop.fs.Path(outputDirectory);
+        final String chunkPath = outputPath + "/" + basename + ".chunk." + chunkNum;
+
+        try (SeekableByteChannel channel = Files.newByteChannel(gcsPath);
+             final OutputStream outputStream = new BufferedOutputStream(BucketUtils.createFile(chunkPath, authHolder))){
+
+            final long start = chunkSize * (long) chunkNum;
+            channel.position(start);
+            final int bufferSize = 512;
```

Comment:
I had no idea how to set the buffer size and was just following online examples which tended to have small buffer sizes. After some experimenting, it looks like setting the buffer size high, and (crucially) using `ByteBuffer.allocateDirect` results in the best performance. There's a lot of variability in timing from run to run as I keep trying it, so I'm just going to set it to the min of 64MB and the chunk size for now.